File: src/bin/executor.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <readline/history.h>
#include <readline/readline.h>

void get_args(char *args[], char scanline[]){

    args[0] = scanline;

    size_t j = 1;
    for(size_t i = 0; i < strlen(scanline); i++){
        if(scanline[i] == ' ') {
            scanline[i] = '\0';
            args[j] = &(scanline[i+1]);
            j++;
        }
    }
    args[j] = NULL;
}



int main(int argc, char *argv[]){

    pid_t pid;
    char *scanline;
    char *mypath;
    char *myprompt;
    char *args[256];
    char path[256];

    mypath = getenv("MYPATH");
    if (mypath == NULL) {
        mypath = "/bin";
    }

    myprompt = getenv("MYPROMPT");
    if ( myprompt == NULL) {
        myprompt = "> ";
    }

    do {
        scanline = readline(myprompt);
        if (scanline == NULL) {
            printf ("exit\n");
            exit (0);
        }

        add_history(scanline);
        get_args(args, scanline);

        if (strcmp (args[0], "exit") == 0) {
            exit (0);
        }

        pid = fork();

        if(pid == 0) {
            sprintf(path , "%s/%s", mypath , args [0]);
            printf ("Hello from Childprocess, start : %s\n", path);
            execv(path, args);
            exit(-1); //only out of an error
        }

        if(pid != 0){
            printf ( "Process PID : % d\n ", getpid());
            printf ("Hello from Mainprocess, the Childprocess is %d.\n", pid);
            wait(NULL); // wait on child
        }
    } while(1);
}

File: src/bin/executor.rs
use std::env;
use std::ffi::{CString};
use std::io::{self, Write};
use libc::{fork, execl, waitpid, c_char, pid_t};

fn split_args(line: &str) -> Vec<String>{
    line.split_whitespace()
        .map(|s| s.to_string())
        .collect()
}


fn main(){
    //read path and prompt and set to default if NULL
    let mypath = env::var("MYPATH").unwrap_or("/bin".to_string());
    let myprompt = env::var("MYPROMPT").unwrap_or("> ".to_string());

    //print path
    print!("{}", myprompt);
    io::stdout().flush().unwrap();

    //read line
    let mut scanline = String::new();

    if io::stdin().read_line(&mut scanline).is_err() {
            println!("exit");
            return;
        }

    //ask againg if its only newline    
    if scanline.trim().is_empty() {
        continue;
    }

    let scanline = scanline.trim();

    //exit command
    if scanline == "exit" {
        return;
    }

    //split args
    let args = split_args(scanline);
    let cmd = &args[0];

    //path to run /bin/cmd
    let fullpath = format!("{}/{}", mypath, cmd);

    //convert to c strings
    let c_path = CString::new(fullpath.clone()).unwrap();

    //execl arguments
    let mut c_args: Vec<CString> = args.iter().map(|a| CString::new(a.as_str()).unwrap()).collect();
    c_args.push(CString::new("").unwrap()); // NULL terminator için

    //wrap them all arguments up to c pointer
    let mut ptrs: Vec<*const c_char> = c_args.iter().map(|s| s.as_ptr()).collect();
    ptrs.push(std::ptr::null());

    unsafe {
        // fork()
        let pid: pid_t = fork();

        if pid == 0 {
            // Child process
            println!("Hello from Childprocess, start: {}", fullpath);
            execl(
                c_path.as_ptr(),
                ptrs[0],
                ptrs[1],
                ptrs[2],
                ptrs[3],
                std::ptr::null::<c_char>(),
            );
            // fail case:
            std::process::exit(-1);
        } else {
            // Parent process
            println!("Hello from Mainprocess, the Childprocess is {}.", pid);
            waitpid(pid, std::ptr::null_mut(), 0);
        }
    }
}
